### PR TITLE
fix(tito): derive suffix ids by token-level slicing to stop boundary drift

### LINKED ISSUE #1319: [BUG] TITO render suffix token id drift
# Miles TITO text-suffix tokenization can diverge from full prompt tokenization

## Summary

Miles `TITOTokenizer` computes appended non-assistant turns by rendering:

1. a synthetic prefix, for example `[dummy_system]` for user/system appends
2. the same synthetic prefix plus appended messages

Then it slices the rendered **textual** suffix and calls `tokenizer.encode(suffix, add_special_tokens=False)`.

Note that miles sclices on text level instead of toke id level. So this actually assumes:

```python
encode(full_text)[len(encode(prefix_text)):] == encode(full_text[len(prefix_text):])
```

That assumption is not always true. With SentencePiece-style tokenizers such as
`HuggingFaceH4/zephyr-7b-beta`, encoding a suffix as an independent text segment
can introduce a beginning-of-word marker token (`▁<`, 523) that is not consistent with
the same character ('<',28789) are tokenized inside the complete message list.

If we use TokenSeqComparator to check this example, the comparator will also report that 
```
ids_equal = false
expected_len = 69
actual_len = 69
mismatch_count = 2
```

## Reproduction

This is tokenizer-only. It does not load model weights and does not require GPU.

Run the following script from the repository root with a Python environment that
has `transformers` available:

<details>
<summary>Reproduction script</summary>

```python
#!/usr/bin/env python3
from __future__ import annotations

import json
import sys
import tempfile
import types
from pathlib import Path
from typing import Any

from transformers import AutoTokenizer


MODEL = "HuggingFaceH4/zephyr-7b-beta"
OUT_DIR = Path(tempfile.gettempdir())


def install_sglang_tool_stub() -> None:
    """Avoid requiring a full sglang install for this tokenizer-only repro."""
    modules = [
        "sglang",
        "sglang.srt",
        "sglang.srt.entrypoints",
        "sglang.srt.entrypoints.openai",
        "sglang.srt.entrypoints.openai.protocol",
    ]
    for name in modules:
        sys.modules.setdefault(name, types.ModuleType(name))
    sys.modules["sglang.srt.entrypoints.openai.protocol"].Tool = object


def normalize_ids(raw: Any) -> list[int]:
    if hasattr(raw, "input_ids"):
        raw = raw.input_ids
    elif isinstance(raw, dict):
        raw = raw["input_ids"]
    if hasattr(raw, "tolist"):
        raw = raw.tolist()
    if raw and isinstance(raw[0], list):
        assert len(raw) == 1
        raw = raw[0]
    return list(raw)


def first_diff(left: list[int], right: list[int]) -> int | None:
    for index, (left_id, right_id) in enumerate(zip(left, right)):
        if left_id != right_id:
            return index
    if len(left) != len(right):
        return min(len(left), len(right))
    return None


def main() -> None:
    install_sglang_tool_stub()
    sys.path.insert(0, str(Path.cwd() / "miles"))

    from miles.utils.chat_template_utils.tito_tokenizer import TITOTokenizer

    tokenizer = AutoTokenizer.from_pretrained(MODEL)
    tito = TITOTokenizer(tokenizer, allowed_append_roles=["user"])

    # Current Miles defines rf_tokenize_user_and_system_segment, while
    # tokenize_additional_non_assistant calls _tokenize_user_and_system_segment.
    # Alias it here so this repro focuses on the suffix tokenization issue.
    if not hasattr(tito, "_tokenize_user_and_system_segment") and hasattr(
        tito, "rf_tokenize_user_and_system_segment"
    ):
        tito._tokenize_user_and_system_segment = tito.rf_tokenize_user_and_system_segment

    # Zephyr's chat template ignores role="tool", so use a supported
    # non-assistant append: system + user + assistant + user.
    old_messages = [
        {"role": "system", "content": "You are a concise assistant."},
        {"role": "user", "content": "Give me a short greeting."},
        {"role": "assistant", "content": "Hello there."},
    ]
    new_messages = old_messages + [
        {"role": "user", "content": "Now answer with one word only."}
    ]

    prefix_text = tito.render_messages(old_messages, add_generation_prompt=False, tools=None)
    prefix_ids = tokenizer.encode(prefix_text, add_special_tokens=False)
    tito_ids = tito.merge_tokens(old_messages, new_messages, prefix_ids, tools=None)

    direct_text = tito.render_messages(new_messages, add_generation_prompt=True, tools=None)
    direct_ids_from_text = tokenizer.encode(direct_text, add_special_tokens=False)
    direct_ids_tokenize_true = normalize_ids(
        tokenizer.apply_chat_template(new_messages, tokenize=True, add_generation_prompt=True)
    )

    diff_index = first_diff(tito_ids, direct_ids_from_text)
    diff_window = None
    if diff_index is not None:
        start = max(0, diff_index - 12)
        end = min(max(len(tito_ids), len(direct_ids_from_text)), diff_index + 24)
        diff_window = {
            "start": start,
            "end": end,
            "tito_ids": tito_ids[start:end],
            "direct_ids": direct_ids_from_text[start:end],
            "tito_tokens": tokenizer.convert_ids_to_tokens(tito_ids[start:end]),
            "direct_tokens": tokenizer.convert_ids_to_tokens(direct_ids_from_text[start:end]),
            "tito_decoded": tokenizer.decode(tito_ids[start:end], skip_special_tokens=False),
            "direct_decoded": tokenizer.decode(direct_ids_from_text[start:end], skip_special_tokens=False),
        }

    summary = {
        "model": MODEL,
        "used_messages_shape": "system + user + assistant + user",
        "reason_for_fallback": "Zephyr chat_template has user/system/assistant branches only; tool role is ignored.",
        "direct_text_encode_equals_tokenize_true": direct_ids_from_text == direct_ids_tokenize_true,
        "tito_equals_direct_text_encode": tito_ids == direct_ids_from_text,
        "tito_equals_direct_tokenize_true": tito_ids == direct_ids_tokenize_true,
        "tito_len": len(tito_ids),
        "direct_len": len(direct_ids_from_text),
        "first_diff_index": diff_index,
        "diff_window": diff_window,
        "tito_decoded": tokenizer.decode(tito_ids, skip_special_tokens=False),
        "direct_decoded": tokenizer.decode(direct_ids_from_text, skip_special_tokens=False),
    }

    (OUT_DIR / "zephyr_miles_tito_ids.json").write_text(json.dumps(tito_ids, ensure_ascii=False, indent=2))
    (OUT_DIR / "zephyr_miles_direct_ids.json").write_text(
        json.dumps(direct_ids_from_text, ensure_ascii=False, indent=2)
    )
    (OUT_DIR / "zephyr_miles_direct_tokenize_true_ids.json").write_text(
        json.dumps(direct_ids_tokenize_true, ensure_ascii=False, indent=2)
    )
    (OUT_DIR / "zephyr_miles_compare.json").write_text(json.dumps(summary, ensure_ascii=False, indent=2))

    print(json.dumps(summary, ensure_ascii=False, indent=2))


if __name__ == "__main__":
    main()
```

</details>

## Observed result

The TITO-assembled sequence and direct full-template sequence have the same
length, but different token IDs:

```text
tito_len = 69
direct_len = 69
tito == direct = false
first_diff_index = 46
```

The first differing token:

```python
Miles TITO: 523   # "▁<"
Direct:     28789 # "<"
```

Diff window:

```text
Miles TITO decoded:
... Hello there.</s>
 <|user|>
Now answer with one word only.</s>
 <|assistant|>

Direct decoded:
... Hello there.</s>
<|user|>
Now answer with one word only.</s>
<|assistant|>
```

The text-level suffix path introduced extra spaces before `<|user|>` and
`<|assistant|>` once the suffix tokens were appended into the existing runtime
token stream.

## Expected result

The TITO-assembled runtime **non-assistant** token sequence should exactly match the canonical full prompt in **token id** level, as indicated by TokenSeqComparator

## Root cause

The current suffix-tokenization primitive is **text**-based:

```python
text_without = render_messages(base_messages, add_generation_prompt=False)
text_with = render_messages(base_messages + appended_messages, add_generation_prompt=...)
suffix_ids = tokenizer.encode(text_with[len(text_without):], add_special_tokens=False)
```

This re-tokenizes the suffix without the original left context. Tokenizers are
not guaranteed to be context-free at arbitrary text boundaries. In Zephyr's
Mistral/LLaMA-style SentencePiece tokenizer, the beginning of an independently
encoded segment is treated like a word boundary. Therefore a leading `<` can be
encoded as `▁<` instead of `<`.

## Possible fix

Avoid slicing at the rendered-text level. Render both the prefix and full prompt
directly to token IDs, then slice at the token-ID level:

```python
def tokenize_rendered_suffix_token_id(
    self,
    base_messages: list[dict],
    appended_messages: list[dict],
    *,
    tools: list[dict] | None = None,
    add_generation_prompt: bool = False,
) -> list[int]:
    prefix_ids = self.render_messages(
        base_messages,
        add_generation_prompt=False,
        tools=tools,
        tokenize=True,
    )
    full_ids = self.render_messages(
        base_messages + appended_messages,
        add_generation_prompt=add_generation_prompt,
        tools=tools,
        tokenize=True,
    )
    if full_ids[: len(prefix_ids)] != prefix_ids:
        raise ValueError("token-id suffix diff failed")
    return full_ids[len(prefix_ids):]
```

Then use this primitive for:

- tool-response segments
- user/system append segments
- final generation-prompt suffix

This makes the incremental tokens match the canonical full prompt tokenization
for tokenizers where suffix-only encoding is not equivalent to full-context
encoding.

## Appendix: Why the example in this issue uses a user incremental msg instead of tool incremental msg

`HuggingFaceH4/zephyr-7b-beta`'s chat template only branches on:

- `user`
- `system`
- `assistant`

It does not render `role="tool"` messages. A tool-message repro is therefore not
meaningful for Zephyr because the tool response is ignored by the template. This
repro uses the supported append shape:

```text
system + user + assistant + user
```

The same text-suffix bug appears at the boundary before the appended `<|user|>`
and again before the final generation prompt `<|assistant|>`.

### PR DESCRIPTION
**Pull request by EazyReal**
## What

`TITOTokenizer._tokenize_rendered_suffix` derived the appended segment's token ids by slicing the rendered **text** (`text_with[len(text_without):]`) and re-encoding that substring on its own. When a token spans the prefix/suffix boundary — a BPE merge across the cut, or a leading-space token formed differently in isolation — the standalone re-encode diverges from the full-prompt tokenization, so the incremental (token-in-token-out) ids drift from what the model actually saw.

## Fix

Encode the base prompt and the full prompt (both `add_special_tokens=False`) and take the suffix as `full_ids[len(prefix_ids):]`. This matches the model's own tokenization and correctly attributes a start-of-segment marker to the full-prompt encoding (the Zephyr `▁<` vs `<` case in #1319).

If `encode(prefix)` is not a token-prefix of `encode(full)` — a junction-spanning merge that would change the already-frozen prefix — raise `ValueError`, matching the existing text-prefix check, instead of silently dropping or duplicating tokens. Trimming the frozen prefix in `merge_tokens` is out of scope.

Tests: start-of-segment drift, a no-op char tokenizer, and the junction-merge guard.

Fixes #1319

### CODE REVIEW

**Comment by EazyReal** [comment]
Thanks — good catch on the boundary case. I switched to the blind token slice `full_ids[len(prefix_ids):]`, which is exactly the value issue #1319 defines as the expected suffix and avoids the duplication you flagged: the earlier common-prefix variant returned the boundary-spanning token, which then duplicated against the independently-encoded prefix in `merge_tokens`. The remaining length-changing boundary-span case (where the *prefix* tokenization itself shifts) is inherent to the incremental `prefix + suffix` design — resolving it would require trimming the prefix inside `merge_tokens`, which is out of scope for this fix.

**Comment by EazyReal** [comment]
@fzyzcjy @yueming-yuan @guapisolo @maocheng23 @jybsuper @Zhichenzzz @Shi-Dong Could I get a quick review on this TITO boundary fix? It derives suffix ids by token slicing so merged-token boundaries do not drift.

**Comment by behroozazarkhalili** [comment]
Verified on top of current main (2ed189c0d); merges with no conflicts.

- tests/fast/utils/chat_template_utils on the merge: 1802 passed, 10 skipped.
- Same suite with tito_tokenizer.py reverted to main: only the new test fails, so it pins the fix.
- ruff and isort clean; black (24.3.0, line length 119) would rewrap the apply_chat_template signature in the new test file, so the pre-commit job will flag that one line.

One suggestion on the slice itself: full_ids[len(prefix_ids):] assumes encode(prefix_text) is a token-prefix of encode(full_text). When a tokenizer merges across the junction, the last prefix token changes and the slice silently drops or duplicates a token, which is the same class of drift #1319 describes from the other side. A one-line guard, `if full_ids[:len(prefix_ids)] != prefix_ids: raise ValueError(...)`, turns that into a loud failure like the existing text-prefix check above it. Worth adding here or as a follow-up.

cc @guapisolo @Zhichenzzz

File: miles/utils/chat_template_utils/tito_tokenizer.py [hunk 1]
```diff
@@ -192,7 +198,14 @@ def _tokenize_rendered_suffix(
...
             raise ValueError(f"rendered suffix diff failed for {roles}")
-        return self._encode_text(text_with[len(text_without) :])
+        prefix_ids = self._encode_text(text_without)
+        full_ids = self._encode_text(text_with)
+        common = 0
+        for prefix_id, full_id in zip(prefix_ids, full_ids, strict=False):
+            if prefix_id != full_id:
+                break
+            common += 1
+        return full_ids[common:]
```
[the remark this answers is not in the document]
Comment (EazyReal) [reply miles/utils/chat_template_utils/tito_tokenizer.py]: Addressed on the current head: suffix ids are `full_ids[len(prefix_ids):]`, not a common-prefix slice, so `merge_tokens` cannot concatenate both the unmerged prefix token and a boundary-spanning suffix token.

When a junction merge would change the prefix encoding itself, the new `token-id suffix diff failed` guard raises instead of slicing. Trimming the frozen prefix is still out of scope for this PR.

**Comment by EazyReal** [comment]
Still needed: `main` still re-encodes the rendered text suffix, and that still diverges on Zephyr (`HuggingFaceH4/zephyr-7b-beta`). On this head, `merge_tokens` matches `encode(full prompt)` (69/69 ids); the old standalone suffix encode still emits `▁<` (523) where the in-context encode has `<` (28789).

Pushed a follow-up that adds the prefix-ids guard suggested above, plus tests for the start-of-segment case, the no-op case, and the loud failure when a junction merge would shift the prefix.

Remaining merge gates I cannot clear from the fork: a `/miles/utils` CODEOWNERS review, and a maintainer `run-ci*` label so GitHub will approve the held workflows.

**Comment by behroozazarkhalili** [comment]
Re-ran on ee959b309 merged onto today's main (df0e677f6): tests/fast/utils/chat_template_utils is 1804 passed, 10 skipped. With tito_tokenizer.py at main, the in-context-ids case and the new prefix-guard case fail (the no-boundary-span case passes either way, as it should). ruff/black/isort clean. The guard reads as I hoped: a junction merge that changes the prefix encoding now raises "token-id suffix diff failed" instead of slicing past it. Nothing left from my side.

cc @guapisolo @Zhichenzzz

**Comment by EazyReal** [comment]
Thanks for the re-check. Agreed: the remaining junction-merge case is a loud failure here (`token-id suffix diff failed`) rather than a silent slice. Trimming the already-frozen prefix in `merge_tokens` is still out of scope.

Still blocked on a `/miles/utils` CODEOWNERS review and a maintainer `run-ci*` label so the held workflows can run.

@guapisolo @Zhichenzzz @fzyzcjy @yueming-yuan
